From 54109b6ae8a117feb21f548394bea90c8b9d8cd9 Mon Sep 17 00:00:00 2001
From: Simon Grund <simon.grund@tuhh.de>
Date: Wed, 4 Feb 2026 13:56:09 +0100
Subject: [PATCH 1/9] Initial port of bplib

Changed all Threading synchronization to RIOT analogs. Changed the time
module to use vfs. The storage is now only the header, the actual
implementation is not fixed inside of bplib but must be supplied by the
user (this could also be a pseudomodule or similar in the package).
Changed other limitations for RIOT (like only one worker).
---
 aa/as/src/bplib_as_internal.c     |  46 ++-----------
 aa/as/src/bplib_as_internal.h     |   6 +-
 aa/nc/inc/bplib_nc_rwlock.h       |   9 +--
 aa/nc/src/bplib_nc.c              |  15 ++---
 aa/nc/src/bplib_nc_rwlock.c       |  56 +++++-----------
 bpa/ct/inc/bplib_ct.h             |   4 +-
 bpa/ct/src/bplib_ct.c             |  36 +++++-----
 bpa/ct/src/bplib_ct_ccs.c         |  16 ++---
 bpa/pi/src/bplib_pi.c             |   2 +-
 bpa/stor/inc/bplib_stor.h         |  12 ++--
 bpa/stor/src/bplib_stor_cmake.c   |   3 +
 ci/mem/inc/bplib_mem.h            |   6 +-
 ci/mem/src/bplib_mem.c            |  13 ++--
 ci/qm/inc/bplib_qm.h              |   2 +-
 ci/qm/inc/bplib_qm_waitqueue.h    |   9 +--
 ci/qm/src/bplib_qm.c              |  10 +--
 ci/qm/src/bplib_qm_waitqueue.c    | 105 ++++++++++--------------------
 ci/time/inc/bplib_time.h          |   2 +
 ci/time/src/bplib_time_internal.c |  62 +++++-------------
 ci/time/src/bplib_time_internal.h |   3 +-
 cla/src/bplib_cla.c               |   2 +-
 inc/bplib_inst.h                  |   3 +-
 version/src/bplib_version.c       |   9 +--
 23 files changed, 155 insertions(+), 276 deletions(-)
 create mode 100644 bpa/stor/src/bplib_stor_cmake.c

diff --git a/aa/as/src/bplib_as_internal.c b/aa/as/src/bplib_as_internal.c
index ffc5f4d2..ca30744b 100644
--- a/aa/as/src/bplib_as_internal.c
+++ b/aa/as/src/bplib_as_internal.c
@@ -38,7 +38,7 @@ BPLib_NodeMibReportsHkTlm_Payload_t    BPLib_AS_NodeReportsPayload;    /** \brie
 /* =============== */
 /* Mutex Variables */
 /* =============== */
-osal_id_t MutexId;
+mutex_t BPLib_AS_Mutex;
 BPLib_TIME_MonotonicTime_t InitTime;    /** \brief Time the node was initialized */
 
 /* ==================== */
@@ -99,55 +99,21 @@ BPLib_Status_t BPLib_AS_SetCounter(BPLib_EID_t EID, BPLib_AS_Counter_t Counter,
 
 BPLib_Status_t BPLib_AS_InitMutex(void)
 {
-    uint32 OS_Status;
-    BPLib_Status_t Status;
-    char MutexName[BPLIB_AS_MAX_MUTEX_NAME_SIZE];
-
-    MutexId = OS_OBJECT_ID_UNDEFINED;
-    strncpy(MutexName, "BPLib_AS_CounterMut", BPLIB_AS_MAX_MUTEX_NAME_SIZE);
-
     /* Instantiate a mutex for AS counters */
-    OS_Status = OS_MutSemCreate(&MutexId, MutexName, 0);
-
-    /* Translate mutex status into BPLib_Status_t */
-    if (OS_Status == OS_SUCCESS)
-    {
-        Status = BPLIB_SUCCESS;
-    }
-    else
-    {
-        Status = BPLIB_AS_INIT_MUTEX_ERR;
-    }
+    mutex_init(&BPLib_AS_Mutex);
 
-    return Status;
+    /* Always return success as this cannot fail in RIOT */
+    return BPLIB_SUCCESS;
 }
 
 void BPLib_AS_LockCounters(void)
 {
-    uint32 OS_Status;
-
-    OS_Status = OS_MutSemTake(MutexId);
-    if (OS_Status != OS_SUCCESS)
-    {
-        BPLib_EM_SendEvent(BPLIB_AS_TAKE_MUTEX_ERR_EID,
-                            BPLib_EM_EventType_ERROR,
-                            "Failed to take from the counter mutex, RC = %d",
-                            OS_Status);
-    }
+    mutex_lock(&BPLib_AS_Mutex);
 }
 
 void BPLib_AS_UnlockCounters(void)
 {
-    uint32 OS_Status;
-
-    OS_Status = OS_MutSemGive(MutexId);
-    if (OS_Status != OS_SUCCESS)
-    {
-        BPLib_EM_SendEvent(BPLIB_AS_GIVE_MUTEX_ERR_EID,
-                            BPLib_EM_EventType_ERROR,
-                            "Failed to give to the counter mutex, RC = %d",
-                            OS_Status);
-    }
+    mutex_unlock(&BPLib_AS_Mutex);
 }
 
 /* Initialize the static MIB reports telemetry values */
diff --git a/aa/as/src/bplib_as_internal.h b/aa/as/src/bplib_as_internal.h
index cb030e5c..683f9f52 100644
--- a/aa/as/src/bplib_as_internal.h
+++ b/aa/as/src/bplib_as_internal.h
@@ -28,7 +28,7 @@
 #include "bplib_api_types.h"
 #include "bplib_as.h"
 #include "bplib_eventids.h"
-#include "osapi.h"
+#include "mutex.h" /* use RIOT mutex directly */
 #include "bplib_eid.h"
 
 /* ======= */
@@ -47,8 +47,8 @@ extern BPLib_NodeMibReportsHkTlm_Payload_t     BPLib_AS_NodeReportsPayload;
 /* =============== */
 /* Mutex Variables */
 /* =============== */
-extern osal_id_t MutexId;
-extern char      MutexName[BPLIB_AS_MAX_MUTEX_NAME_SIZE];
+/* I dont think this name as a global is very good, so I prefixed it */
+extern mutex_t  BPLib_AS_Mutex;
 
 /* =================== */
 /* Function Prototypes */
diff --git a/aa/nc/inc/bplib_nc_rwlock.h b/aa/nc/inc/bplib_nc_rwlock.h
index 2fe65fff..3e5c7a19 100644
--- a/aa/nc/inc/bplib_nc_rwlock.h
+++ b/aa/nc/inc/bplib_nc_rwlock.h
@@ -22,7 +22,8 @@
 
 #include "bplib_api_types.h"
 
-#include <pthread.h>
+#include "mutex.h"
+#include "cond.h"
 
 /**
  * \brief Definition of the read-write lock structure.
@@ -32,9 +33,9 @@
  * for readers and writers, and counters for the number of readers and writers.
  */
 typedef struct BPLib_NC_RWLock {
-    pthread_mutex_t Lock;          /**< Mutex to protect shared data */
-    pthread_cond_t ReadCond;       /**< Condition variable for readers */
-    pthread_cond_t WriteCond;      /**< Condition variable for writers */
+    mutex_t Lock;          /**< Mutex to protect shared data */
+    cond_t ReadCond;       /**< Condition variable for readers */
+    cond_t WriteCond;      /**< Condition variable for writers */
     int ReaderCnt;                 /**< Number of active readers */
     int WriterCnt;               /**< Number of active writers */
 } BPLib_NC_RWLock_t;
diff --git a/aa/nc/src/bplib_nc.c b/aa/nc/src/bplib_nc.c
index 8eea6f34..ba533f6b 100644
--- a/aa/nc/src/bplib_nc.c
+++ b/aa/nc/src/bplib_nc.c
@@ -147,14 +147,13 @@ BPLib_Status_t BPLib_NC_Init(BPLib_NC_ConfigPtrs_t* ConfigPtrs, void* Callbacks,
     /* Without modifying the TIME module, this was unable to work in BPCat. The
     ** TIME and AS module's use of OSAL needs to be abstracted because no data is
     ** being put into BPLib by BPCat */
-    #ifdef CFS_BUILD
-        /* Initialize time */
-        Status = BPLib_TIME_Init();
-        if (Status != BPLIB_SUCCESS)
-        { /* Failed initialization of TIME */
-            return BPLIB_NC_TIME_INIT_ERR;
-        }
-    #endif
+    /* This should work on RIOT */
+    /* Initialize time */
+    Status = BPLib_TIME_Init();
+    if (Status != BPLIB_SUCCESS)
+    { /* Failed initialization of TIME */
+        return BPLIB_NC_TIME_INIT_ERR;
+    }
 
     /* Initialize NC */
     Status = BPLib_NC_InitImpl(Instance, ConfigPtrs);
diff --git a/aa/nc/src/bplib_nc_rwlock.c b/aa/nc/src/bplib_nc_rwlock.c
index fa8b3533..309defde 100644
--- a/aa/nc/src/bplib_nc_rwlock.c
+++ b/aa/nc/src/bplib_nc_rwlock.c
@@ -27,23 +27,9 @@ BPLib_Status_t BPLib_NC_RWLock_Init(BPLib_NC_RWLock_t *RWLock)
         return BPLIB_NULL_PTR_ERROR;
     }
 
-    if (pthread_mutex_init(&RWLock->Lock, NULL) != 0)
-    {
-        return BPLIB_OS_ERROR;
-    }
-
-    if (pthread_cond_init(&RWLock->ReadCond, NULL) != 0)
-    {
-        pthread_mutex_destroy(&RWLock->Lock);
-        return BPLIB_OS_ERROR;
-    }
-
-    if (pthread_cond_init(&RWLock->WriteCond, NULL) != 0)
-    {
-        pthread_mutex_destroy(&RWLock->Lock);
-        pthread_cond_destroy(&RWLock->ReadCond);
-        return BPLIB_OS_ERROR;
-    }
+    mutex_init(&RWLock->Lock);
+    cond_init(&RWLock->ReadCond);
+    cond_init(&RWLock->WriteCond);
 
     RWLock->ReaderCnt = 0;
     RWLock->WriterCnt = 0;
@@ -53,14 +39,8 @@ BPLib_Status_t BPLib_NC_RWLock_Init(BPLib_NC_RWLock_t *RWLock)
 
 void BPLib_NC_RWLock_Destroy(BPLib_NC_RWLock_t *RWLock)
 {
-    if (RWLock == NULL)
-    {
-        return;
-    }
-
-    pthread_mutex_destroy(&RWLock->Lock);
-    pthread_cond_destroy(&RWLock->ReadCond);
-    pthread_cond_destroy(&RWLock->WriteCond);
+    // No mutex cleanup in RIOT
+    return;
 }
 
 void BPLib_NC_RWLock_RLock(BPLib_NC_RWLock_t *RWLock)
@@ -70,16 +50,16 @@ void BPLib_NC_RWLock_RLock(BPLib_NC_RWLock_t *RWLock)
         return;
     }
 
-    pthread_mutex_lock(&RWLock->Lock);
+    mutex_lock(&RWLock->Lock);
 
     while (RWLock->WriterCnt > 0)
     {
-        pthread_cond_wait(&RWLock->ReadCond, &RWLock->Lock);
+        cond_wait(&RWLock->ReadCond, &RWLock->Lock);
     }
 
     RWLock->ReaderCnt++;
 
-    pthread_mutex_unlock(&RWLock->Lock);
+    mutex_unlock(&RWLock->Lock);
 }
 
 void BPLib_NC_RWLock_RUnlock(BPLib_NC_RWLock_t *RWLock)
@@ -89,16 +69,16 @@ void BPLib_NC_RWLock_RUnlock(BPLib_NC_RWLock_t *RWLock)
         return;
     }
 
-    pthread_mutex_lock(&RWLock->Lock);
+    mutex_lock(&RWLock->Lock);
 
     RWLock->ReaderCnt--;
 
     if (RWLock->ReaderCnt == 0)
     {
-        pthread_cond_signal(&RWLock->WriteCond);
+        cond_signal(&RWLock->WriteCond);
     }
 
-    pthread_mutex_unlock(&RWLock->Lock);
+    mutex_unlock(&RWLock->Lock);
 }
 
 void BPLib_NC_RWLock_WLock(BPLib_NC_RWLock_t *RWLock)
@@ -108,16 +88,16 @@ void BPLib_NC_RWLock_WLock(BPLib_NC_RWLock_t *RWLock)
         return;
     }
 
-    pthread_mutex_lock(&RWLock->Lock);
+    mutex_lock(&RWLock->Lock);
 
     while (RWLock->ReaderCnt > 0 || RWLock->WriterCnt > 0)
     {
-        pthread_cond_wait(&RWLock->WriteCond, &RWLock->Lock);
+        cond_wait(&RWLock->WriteCond, &RWLock->Lock);
     }
 
     RWLock->WriterCnt++;
 
-    pthread_mutex_unlock(&RWLock->Lock);
+    mutex_unlock(&RWLock->Lock);
 }
 
 void BPLib_NC_RWLock_WUnlock(BPLib_NC_RWLock_t *RWLock)
@@ -127,12 +107,12 @@ void BPLib_NC_RWLock_WUnlock(BPLib_NC_RWLock_t *RWLock)
         return;
     }
 
-    pthread_mutex_lock(&RWLock->Lock);
+    mutex_lock(&RWLock->Lock);
 
     RWLock->WriterCnt--;
 
-    pthread_cond_broadcast(&RWLock->ReadCond);
-    pthread_cond_signal(&RWLock->WriteCond);
+    cond_broadcast(&RWLock->ReadCond);
+    cond_signal(&RWLock->WriteCond);
 
-    pthread_mutex_unlock(&RWLock->Lock);
+    mutex_unlock(&RWLock->Lock);
 }
diff --git a/bpa/ct/inc/bplib_ct.h b/bpa/ct/inc/bplib_ct.h
index c8302eee..e3d1c199 100644
--- a/bpa/ct/inc/bplib_ct.h
+++ b/bpa/ct/inc/bplib_ct.h
@@ -28,7 +28,7 @@
 #include "bplib_api_types.h"
 #include "bplib_rbt.h"
 #include "bplib_bblocks.h"
-#include <pthread.h>
+#include "mutex.h"
 #include "bplib_cfg.h"
 
 /*
@@ -215,7 +215,7 @@ typedef struct
     BPLib_RBT_Root_t SeqTreeRoot;                               /** \brief An RBT that allows for CTDB queries based on sequence ID/number */
     BPLib_RBT_Root_t IdTreeRoot;                                /** \brief An RBT that allows for CTDB queries based on bundle ID */
 
-    pthread_mutex_t Lock;                                     /** \brief Read/write lock on CTDB and open CCSs */
+    mutex_t Lock;                                               /** \brief Read/write lock on CTDB and open CCSs */
 } BPLib_CT_Context_t;
 
 /*
diff --git a/bpa/ct/src/bplib_ct.c b/bpa/ct/src/bplib_ct.c
index dad95607..ecad2366 100644
--- a/bpa/ct/src/bplib_ct.c
+++ b/bpa/ct/src/bplib_ct.c
@@ -127,7 +127,7 @@ BPLib_Status_t BPLib_CT_Init(BPLib_Instance_t *Inst)
 
     memset(&(Inst->Ct), 0, sizeof(BPLib_CT_Context_t));
 
-    pthread_mutex_init(&Inst->Ct.Lock, NULL);
+    mutex_init(&Inst->Ct.Lock);
 
     BPLib_RBT_InitRoot(&(Inst->Ct.SeqTreeRoot));
     BPLib_RBT_InitRoot(&(Inst->Ct.IdTreeRoot));
@@ -197,7 +197,7 @@ BPLib_Status_t BPLib_CT_ProcessNewBundle(BPLib_Instance_t* Inst, BPLib_Bundle_t
     /* Default to refused custody */
     DispCode = BPLib_CT_CustodyRefused;
 
-    pthread_mutex_lock(&Inst->Ct.Lock);
+    mutex_lock(&Inst->Ct.Lock);
 
     /* Reject custody due to lack of storage */
     /* Check if there's storage left (TotalBytes includes the deserialized bundle) */
@@ -257,7 +257,7 @@ BPLib_Status_t BPLib_CT_ProcessNewBundle(BPLib_Instance_t* Inst, BPLib_Bundle_t
         (void) BPLib_CT_SignalCustodyImpl(Inst, Bundle, DispCode, StoreBundle, ExtBlockIdx);
     }    
     
-    pthread_mutex_unlock(&Inst->Ct.Lock);
+    mutex_unlock(&Inst->Ct.Lock);
 
     return Status;
 }
@@ -280,10 +280,10 @@ BPLib_Status_t BPLib_CT_SignalCustody(BPLib_Instance_t *Inst, BPLib_Bundle_t *Bu
         return BPLIB_SUCCESS;
     }
 
-    pthread_mutex_lock(&Inst->Ct.Lock);
+    mutex_lock(&Inst->Ct.Lock);
     Status = BPLib_CT_SignalCustodyImpl(Inst, Bundle, DispCode, StoreBundle, ExtBlockIdx);
 
-    pthread_mutex_unlock(&Inst->Ct.Lock);
+    mutex_unlock(&Inst->Ct.Lock);
 
     return Status;
 }
@@ -317,7 +317,7 @@ BPLib_Status_t BPLib_CT_UpdateBundle(BPLib_Instance_t* Inst, BPLib_Bundle_t *Bun
 
     CtebPtr = &(Bundle->blocks.ExtBlocks[ExtBlockIdx].BlockData.CustodyBlockData);
 
-    pthread_mutex_lock(&Inst->Ct.Lock);
+    mutex_lock(&Inst->Ct.Lock);
     
     Status = BPLib_CT_GetEntryFromCtdbWithId(&(Inst->Ct),
                         Bundle->blocks.PrimaryBlock.BundleId, &DbEntry);
@@ -382,7 +382,7 @@ BPLib_Status_t BPLib_CT_UpdateBundle(BPLib_Instance_t* Inst, BPLib_Bundle_t *Bun
         Bundle->blocks.ExtBlocks[ExtBlockIdx].Header.RequiresEncode = true;
     }
 
-    pthread_mutex_unlock(&Inst->Ct.Lock);
+    mutex_unlock(&Inst->Ct.Lock);
 
     return Status;
 }
@@ -422,7 +422,7 @@ BPLib_Status_t BPLib_CT_ProcessCcs(BPLib_Instance_t *Inst, BPLib_CT_Deserialized
         }
     }
 
-    pthread_mutex_unlock(&Inst->Ct.Lock);
+    mutex_unlock(&Inst->Ct.Lock);
 
     return Status;
 }
@@ -463,7 +463,7 @@ void BPLib_CT_DeleteBundleFromCtdb(BPLib_Instance_t *Inst, uint32_t BundleId)
         return;
     }
 
-    pthread_mutex_lock(&Inst->Ct.Lock);
+    mutex_lock(&Inst->Ct.Lock);
 
     Status = BPLib_CT_GetEntryFromCtdbWithId(&(Inst->Ct), BundleId, &DbEntry);
     if (Status == BPLIB_SUCCESS)
@@ -478,14 +478,14 @@ void BPLib_CT_DeleteBundleFromCtdb(BPLib_Instance_t *Inst, uint32_t BundleId)
                     BundleId, Status);
     }
 
-    pthread_mutex_unlock(&Inst->Ct.Lock);
+    mutex_unlock(&Inst->Ct.Lock);
 }
 
 void BPLib_CT_BuildAndSendOpenCcs(BPLib_Instance_t* Instance, BPLib_CT_OpenCcs_t* OpenCcs)
 {
-    pthread_mutex_lock(&Instance->Ct.Lock);
+    mutex_lock(&Instance->Ct.Lock);
     BPLib_CT_BuildAndSendOpenCcs_Impl(Instance, OpenCcs);
-    pthread_mutex_unlock(&Instance->Ct.Lock);
+    mutex_unlock(&Instance->Ct.Lock);
 }
 
 void BPLib_CT_CheckCcsTimeout(BPLib_Instance_t* Instance)
@@ -495,7 +495,7 @@ void BPLib_CT_CheckCcsTimeout(BPLib_Instance_t* Instance)
     int64_t             TimeOpen;
     BPLib_CT_OpenCcs_t* OpenCcs;
 
-    pthread_mutex_lock(&Instance->Ct.Lock);
+    mutex_lock(&Instance->Ct.Lock);
 
     Context = &(Instance->Ct);
     for (OpenCcsIdx = 0; OpenCcsIdx < BPLIB_CT_MAX_OPEN_CCS; OpenCcsIdx++)
@@ -515,7 +515,7 @@ void BPLib_CT_CheckCcsTimeout(BPLib_Instance_t* Instance)
         }
     }
 
-    pthread_mutex_unlock(&Instance->Ct.Lock);
+    mutex_unlock(&Instance->Ct.Lock);
 }
 
 BPLib_Status_t BPLib_CT_CompleteDelivery(BPLib_Instance_t *Inst, BPLib_Bundle_t *Bundle)
@@ -536,7 +536,7 @@ BPLib_Status_t BPLib_CT_CompleteDelivery(BPLib_Instance_t *Inst, BPLib_Bundle_t
         return BPLIB_SUCCESS;
     }
 
-    pthread_mutex_lock(&Inst->Ct.Lock);
+    mutex_lock(&Inst->Ct.Lock);
 
     /* Mark bundle as transferred */
     Status = BPLib_CT_GetEntryFromCtdbWithId(&(Inst->Ct), 
@@ -561,7 +561,7 @@ BPLib_Status_t BPLib_CT_CompleteDelivery(BPLib_Instance_t *Inst, BPLib_Bundle_t
         Status = BPLIB_SUCCESS;
     }
 
-    pthread_mutex_unlock(&Inst->Ct.Lock);
+    mutex_unlock(&Inst->Ct.Lock);
 
     return Status;
 }
@@ -573,7 +573,7 @@ bool BPLib_CT_TriggerCustodialGarbageCollection(BPLib_Instance_t *Inst)
     
     if (Inst != NULL)
     {
-        pthread_mutex_lock(&Inst->Ct.Lock);
+        mutex_lock(&Inst->Ct.Lock);
 
         if (Inst->Ct.CurrDbSize == 0)
         {
@@ -586,7 +586,7 @@ bool BPLib_CT_TriggerCustodialGarbageCollection(BPLib_Instance_t *Inst)
             TriggerGC = (BPLIB_CT_DB_MAX_PERCENT_NONCUSTODIAL_ENTRIES <= (uint32_t)(DecimalVal * 100));
         }
 
-        pthread_mutex_unlock(&Inst->Ct.Lock);
+        mutex_unlock(&Inst->Ct.Lock);
     }
 
     return TriggerGC;
diff --git a/bpa/ct/src/bplib_ct_ccs.c b/bpa/ct/src/bplib_ct_ccs.c
index a7505d2c..d19ed1a3 100644
--- a/bpa/ct/src/bplib_ct_ccs.c
+++ b/bpa/ct/src/bplib_ct_ccs.c
@@ -389,7 +389,7 @@ BPLib_Status_t BPLib_CT_ProcessBundleSeqCollection(BPLib_Instance_t *Inst,
     ** locks at the same time introduces the potential for deadlocks. Each thread should
     ** only ever have one of each lock at a time.
     */
-    pthread_mutex_lock(&Inst->Ct.Lock);
+    mutex_lock(&Inst->Ct.Lock);
 
     for (SeqRangeIdx = 0; SeqRangeIdx < SeqCollection->SeqRangeLen; SeqRangeIdx++)
     {
@@ -431,9 +431,9 @@ BPLib_Status_t BPLib_CT_ProcessBundleSeqCollection(BPLib_Instance_t *Inst,
                     DbEntry->State = BPLib_CT_Transferred;
 
                     /* Request bundle deletion from storage */
-                    pthread_mutex_unlock(&Inst->Ct.Lock);
+                    mutex_unlock(&Inst->Ct.Lock);
                     BPLib_STOR_AddToCustodialUpdateBatch(Inst, BundleId, BPLIB_CT_MARK_DELETE);
-                    pthread_mutex_lock(&Inst->Ct.Lock);
+                    mutex_lock(&Inst->Ct.Lock);
 
                     BPLib_AS_Increment(BPLIB_EID_INSTANCE, BUNDLE_COUNT_CUSTODY_TRANSFERRED, 1);
                 }
@@ -441,9 +441,9 @@ BPLib_Status_t BPLib_CT_ProcessBundleSeqCollection(BPLib_Instance_t *Inst,
                 else
                 {
                     /* Request storage turn retransmit timer off */
-                    pthread_mutex_unlock(&Inst->Ct.Lock);
+                    mutex_unlock(&Inst->Ct.Lock);
                     BPLib_STOR_AddToCustodialUpdateBatch(Inst, BundleId, BPLIB_CT_STOP_RETRANSMIT);
-                    pthread_mutex_lock(&Inst->Ct.Lock);
+                    mutex_lock(&Inst->Ct.Lock);
 
                     BPLib_AS_Increment(BPLIB_EID_INSTANCE, BUNDLE_COUNT_CUSTODY_REJECTED, 1);
                 }
@@ -454,9 +454,9 @@ BPLib_Status_t BPLib_CT_ProcessBundleSeqCollection(BPLib_Instance_t *Inst,
                 BundleId = DbEntry->BundleId;
 
                 /* Request bundle retransmission from storage */
-                pthread_mutex_unlock(&Inst->Ct.Lock);
+                mutex_unlock(&Inst->Ct.Lock);
                 BPLib_STOR_AddToCustodialUpdateBatch(Inst, BundleId, BPLIB_CT_START_RETRANSMIT);
-                pthread_mutex_lock(&Inst->Ct.Lock);
+                mutex_lock(&Inst->Ct.Lock);
             }
         }
 
@@ -488,7 +488,7 @@ BPLib_Status_t BPLib_CT_ProcessBundleSeqCollection(BPLib_Instance_t *Inst,
         CurrSeqNum += SeqCollection->SeqRange[SeqRangeIdx];
     }
 
-    pthread_mutex_unlock(&Inst->Ct.Lock);
+    mutex_unlock(&Inst->Ct.Lock);
 
     /* Do remaining batch of storage operations */
     BPLib_STOR_UpdateCustodialBundles(Inst);
diff --git a/bpa/pi/src/bplib_pi.c b/bpa/pi/src/bplib_pi.c
index b994e655..544415d7 100644
--- a/bpa/pi/src/bplib_pi.c
+++ b/bpa/pi/src/bplib_pi.c
@@ -299,7 +299,7 @@ BPLib_Status_t BPLib_PI_RemoveApplication(BPLib_Instance_t *Inst, uint32_t ChanI
     }
 
     /* Clear relevant load batch. Ignore return code since we've already done a null check */
-    (void) BPLib_STOR_LoadBatch_Reset(&Inst->BundleStorage.ChannelLoadBatches[ChanId]);
+    //(void) BPLib_STOR_LoadBatch_Reset(&Inst->BundleStorage.ChannelLoadBatches[ChanId]);
 
     /* Reset sequence number */
     Inst->ChanCtxt[ChanId].SequenceNum = 0;
diff --git a/bpa/stor/inc/bplib_stor.h b/bpa/stor/inc/bplib_stor.h
index 52f30e5d..93a39c3c 100644
--- a/bpa/stor/inc/bplib_stor.h
+++ b/bpa/stor/inc/bplib_stor.h
@@ -29,8 +29,6 @@
 #include "bplib_cfg.h"
 #include "bplib_eid.h"
 #include "bplib_mem.h"
-#include "bplib_stor_loadbatch.h"
-#include "bplib_stor_sql.h"
 #include "bplib_ct.h"
 
 /* ====== */
@@ -42,7 +40,7 @@
 #endif
 
 #ifndef BPLIB_STOR_DISCARDBATCHSIZE
-#define BPLIB_STOR_DISCARDBATCHSIZE 25000
+#define BPLIB_STOR_DISCARDBATCHSIZE 64
 #endif
 
 /* We conditionally allow this to be defined by a compile time variable
@@ -76,13 +74,13 @@ typedef struct
 
 struct BPLib_BundleCache
 {
-    pthread_mutex_t        lock;
-    sqlite3*               db;
+    mutex_t lock;
     int64_t                BundleRowsToDelete[BPLIB_STOR_DISCARDBATCHSIZE];
     BPLib_Bundle_t*        InsertBatch[BPLIB_STOR_INSERTBATCHSIZE];
     size_t                 InsertBatchSize;
-    BPLib_STOR_LoadBatch_t ChannelLoadBatches[BPLIB_MAX_NUM_CHANNELS];
-    BPLib_STOR_LoadBatch_t ContactLoadBatches[BPLIB_MAX_NUM_CONTACTS];
+    /* The load batches are removed for RIOT because they are very implementation dependent.
+     * Loading the numbers (which iirc were the row numbers in the blob table) does not really
+     * help if the actual storage implementation uses some other indexing / storage medium. */
     BPLib_STOR_CtUpdateBatch_t CustodyUpdateBatch;
     int64_t                LastActiveTime;
     size_t                 BundleCountNotEgressed;  /* Number of bundles in storage that have not been marked as egressed */
diff --git a/bpa/stor/src/bplib_stor_cmake.c b/bpa/stor/src/bplib_stor_cmake.c
new file mode 100644
index 00000000..e4d36e61
--- /dev/null
+++ b/bpa/stor/src/bplib_stor_cmake.c
@@ -0,0 +1,3 @@
+void BPLib_STOR_CMakeIsHappyNow(void) {
+    return;
+}
\ No newline at end of file
diff --git a/ci/mem/inc/bplib_mem.h b/ci/mem/inc/bplib_mem.h
index 78c4594f..bd40125c 100644
--- a/ci/mem/inc/bplib_mem.h
+++ b/ci/mem/inc/bplib_mem.h
@@ -27,7 +27,7 @@
 #include "bplib_ct.h"
 #include "bplib_arp.h"
 
-#include <pthread.h>
+#include "mutex.h"
 
 
 /*
@@ -125,8 +125,8 @@ struct BPLib_MEM_Block
 typedef struct BPLib_MEM_Pool
 {
     BPLib_MEM_PoolImpl_t impl; /**< The pool implementation (details hidden) */
-    pthread_mutex_t lock; /**< Mutex for synchronizing access to the pool */
-    size_t         HighWaterMark;   /** \brief Highwater memory mark */
+    mutex_t lock; /**< Mutex for synchronizing access to the pool */
+    size_t HighWaterMark;   /** \brief Highwater memory mark */
 } BPLib_MEM_Pool_t;
 
 
diff --git a/ci/mem/src/bplib_mem.c b/ci/mem/src/bplib_mem.c
index 80a00c4a..1ae1f826 100644
--- a/ci/mem/src/bplib_mem.c
+++ b/ci/mem/src/bplib_mem.c
@@ -56,7 +56,7 @@ BPLib_Status_t BPLib_MEM_PoolInit(BPLib_MEM_Pool_t* pool, void* init_mem, size_t
     memset(pool, 0, sizeof(BPLib_MEM_Pool_t));
     memset(init_mem, 0, sizeof(init_size));
 
-    pthread_mutex_init(&pool->lock, NULL);
+    mutex_init(&pool->lock);
     
     return BPLib_MEM_PoolImplInit(&pool->impl, init_mem, init_size,
         sizeof(BPLib_MEM_Block_t));
@@ -69,7 +69,6 @@ void BPLib_MEM_PoolDestroy(BPLib_MEM_Pool_t* pool)
         return;
     }
 
-    pthread_mutex_destroy(&pool->lock);
     BPLib_MEM_PoolImplDestroy(&pool->impl);
     memset(pool, 0, sizeof(BPLib_MEM_Pool_t));
 }
@@ -83,10 +82,8 @@ BPLib_MEM_Block_t* BPLib_MEM_BlockAlloc(BPLib_MEM_Pool_t* pool, size_t Size)
         return NULL;
     }
 
-    pthread_mutex_lock(&pool->lock);
-    
+    mutex_lock(&pool->lock);
     block = (BPLib_MEM_Block_t*)(BPLib_MEM_PoolImplAlloc(&pool->impl, Size));
-
     if (block != NULL)
     {
         block->used_len = 0;
@@ -98,7 +95,7 @@ BPLib_MEM_Block_t* BPLib_MEM_BlockAlloc(BPLib_MEM_Pool_t* pool, size_t Size)
         }
     }
 
-    pthread_mutex_unlock(&pool->lock);
+    mutex_unlock(&pool->lock);
 
     return block;
 }
@@ -110,9 +107,9 @@ void BPLib_MEM_BlockFree(BPLib_MEM_Pool_t* pool, BPLib_MEM_Block_t* block)
         return;
     }
 
-    pthread_mutex_lock(&pool->lock);
+    mutex_lock(&pool->lock);
     BPLib_MEM_PoolImplFree(&pool->impl, (void*)block);
-    pthread_mutex_unlock(&pool->lock);
+    mutex_unlock(&pool->lock);
 }
 
 BPLib_MEM_Block_t* BPLib_MEM_BlockListAlloc(BPLib_MEM_Pool_t* pool, size_t byte_len, size_t BlockSize)
diff --git a/ci/qm/inc/bplib_qm.h b/ci/qm/inc/bplib_qm.h
index 2803a190..562c2f2d 100644
--- a/ci/qm/inc/bplib_qm.h
+++ b/ci/qm/inc/bplib_qm.h
@@ -28,7 +28,7 @@
 #define QM_STANDARD_WAIT    100L /**< Standard wait of 100msec */
 #define QM_NO_WAIT          0L   /**< Constant representing no wait */
 #define QM_WAIT_FOREVER    -1L   /**< Constant representing an indefinite wait */
-#define QM_MAX_GEN_WORKERS  8L   /**< Constant representing maximum allowed generic workers */
+#define QM_MAX_GEN_WORKERS  1L   /**< Constant representing maximum allowed generic workers */
 
 typedef struct BPLib_QM_WorkerState
 {
diff --git a/ci/qm/inc/bplib_qm_waitqueue.h b/ci/qm/inc/bplib_qm_waitqueue.h
index 0934f083..62d65ba1 100644
--- a/ci/qm/inc/bplib_qm_waitqueue.h
+++ b/ci/qm/inc/bplib_qm_waitqueue.h
@@ -29,7 +29,8 @@
 #include <stddef.h>
 
 // TODO: Use bplib_os
-#include <pthread.h>
+#include "mutex.h"
+#include "sema.h"
 
 /**
  * @struct BPLib_QM_WaitQueue
@@ -48,9 +49,9 @@ typedef struct BPLib_QM_WaitQueue
     int rear; /**< Index of the rear of the queue */
     size_t size; /**< Current number of elements in the queue */
     
-    pthread_mutex_t lock; /**< Mutex for thread synchronization */
-    pthread_cond_t cv_pull; /**< Condition variable for waiting on pulls */
-    pthread_cond_t cv_push; /**< Condition variable for waiting on pushes */
+    mutex_t lock; /**< Mutex for thread synchronization */
+    sema_t s_items; /**< Condition variable for waiting on pulls -> now a sema of queue items */
+    sema_t s_slots; /**< Condition variable for waiting on pushes -> now a sema of free capacity */
 } BPLib_QM_WaitQueue_t;
 
 /**
diff --git a/ci/qm/src/bplib_qm.c b/ci/qm/src/bplib_qm.c
index 9f2252ea..8cc9f531 100644
--- a/ci/qm/src/bplib_qm.c
+++ b/ci/qm/src/bplib_qm.c
@@ -42,10 +42,7 @@ BPLib_Status_t BPLib_QM_QueueTableInit(BPLib_Instance_t* inst, size_t MaxJobs)
     }
 
     /* Initialize worker registration state */
-    if (pthread_mutex_init(&inst->RegisteredWorkersLock, NULL) != 0)
-    {
-        return BPLIB_QM_INIT_ERROR;
-    }
+    mutex_init(&inst->RegisteredWorkersLock);
 
     /* Init Cache */
     Status = BPLib_STOR_Init(inst);
@@ -112,7 +109,6 @@ void BPLib_QM_QueueTableDestroy(BPLib_Instance_t* inst)
     }
 
     /* Worker State Cleanup */
-    pthread_mutex_destroy(&inst->RegisteredWorkersLock);
     inst->NumWorkers = 0;
 
     /* Queue Cleanup */
@@ -143,7 +139,7 @@ BPLib_Status_t BPLib_QM_RegisterWorker(BPLib_Instance_t* inst, int32_t* WorkerID
         return BPLIB_NULL_PTR_ERROR;
     }
 
-    pthread_mutex_lock(&inst->RegisteredWorkersLock);
+    mutex_lock(&inst->RegisteredWorkersLock);
     if (inst->NumWorkers == QM_MAX_GEN_WORKERS)
     {
         *WorkerID = -1;
@@ -156,7 +152,7 @@ BPLib_Status_t BPLib_QM_RegisterWorker(BPLib_Instance_t* inst, int32_t* WorkerID
         *WorkerID = NewWorkerID;
         Status = BPLIB_SUCCESS;
     }
-    pthread_mutex_unlock(&inst->RegisteredWorkersLock);
+    mutex_unlock(&inst->RegisteredWorkersLock);
 
     return Status;
 }
diff --git a/ci/qm/src/bplib_qm_waitqueue.c b/ci/qm/src/bplib_qm_waitqueue.c
index bb5061ef..5fb883b7 100644
--- a/ci/qm/src/bplib_qm_waitqueue.c
+++ b/ci/qm/src/bplib_qm_waitqueue.c
@@ -26,27 +26,8 @@
 
 #include <errno.h>
 #include <stdio.h>
-#include "osapi.h"
 
-static void ms_to_abstimeout(uint32_t ms, struct timespec *ts)
-{
-    int64_t AbsTimeoutInMsec;
-    OS_time_t time;
-
-    if (ts == NULL)
-    {
-        return;
-    }
-
-    /* This is super cleugy but it gets rid of our clock_gettime dependency for now */
-    OS_GetLocalTime(&time);
-    AbsTimeoutInMsec = (time.ticks / OS_TIME_TICKS_PER_MSEC) + ms;
-
-    ts->tv_sec = AbsTimeoutInMsec / 1000;
-    ts->tv_nsec = (AbsTimeoutInMsec % 1000 * 1000000);
-    ts->tv_sec += ts->tv_nsec / 1000000000;
-    ts->tv_nsec %= 1000000000;
-}
+#include "ztimer.h"
 
 /*******************************************************************************
 * Exported Functions
@@ -72,9 +53,9 @@ bool BPLib_QM_WaitQueueInit(BPLib_QM_WaitQueue_t* q, size_t el_size, size_t capa
     }
 
     // TODO: move to bplib_OS module
-    pthread_mutex_init(&q->lock, NULL);
-    pthread_cond_init(&q->cv_push, NULL);
-    pthread_cond_init(&q->cv_pull, NULL);
+    mutex_init(&q->lock);
+    sema_create(&q->s_slots, capacity);
+    sema_create(&q->s_items, 0);
     return true;
 }
 
@@ -94,14 +75,13 @@ void BPLib_QM_WaitQueueDestroy(BPLib_QM_WaitQueue_t* q)
     q->size = 0;
 
     // TODO: move to bplib_OS module
-    pthread_mutex_destroy(&q->lock);
-    pthread_cond_destroy(&q->cv_push);
-    pthread_cond_destroy(&q->cv_pull);
+    // No mutex cleanup in RIOT
+    sema_destroy(&q->s_slots);
+    sema_destroy(&q->s_items);
 }
 
 bool BPLib_QM_WaitQueueTryPush(BPLib_QM_WaitQueue_t* q, const void* item, int timeout_ms)
 {
-    struct timespec deadline;
     int rc;
 
     if ((q == NULL) || (item == NULL))
@@ -109,78 +89,63 @@ bool BPLib_QM_WaitQueueTryPush(BPLib_QM_WaitQueue_t* q, const void* item, int ti
         return false;
     }
 
-    ms_to_abstimeout((uint32_t)(timeout_ms), &deadline);
-    pthread_mutex_lock(&q->lock);
-    /**** Critical Section Begin ****/
-
     /* Wait for queue to be non-full */
-    while (q->size == q->capacity)
-    {
-        rc = pthread_cond_timedwait(&q->cv_push, &q->lock, &deadline);
-        if (rc != 0)
-        {
-            if (rc != ETIMEDOUT)
-            {
-                printf(" BPLib_QM_WaitQueueTryPush NON-TIMEOUT ERROR: %s\n", strerror(rc));
-            }
-            pthread_mutex_unlock(&q->lock);
-            return false;
+    rc = sema_wait_timed_ztimer(&q->s_slots, ZTIMER_MSEC, timeout_ms);
+    if (rc != 0) {
+        if (rc != -ETIMEDOUT && rc != -EAGAIN) {
+            printf(" BPLib_QM_WaitQueueTryPush NON-TIMEOUT ERROR: %s\n", strerror(rc));
         }
+        return false;
     }
 
+    mutex_lock(&q->lock);
+    /**** Critical Section Begin ****/
+
     /* Push an item */
     q->rear = (q->rear  + 1) % q->capacity;
     memcpy((void*)(((char *)q->storage) + (q->rear*q->el_size)), item, q->el_size);
     q->size++;
 
-    /* Notify other pulling threads that an item can be pulled. */
-    pthread_cond_signal(&q->cv_pull);
-
     /**** Critical Section End ****/
-    pthread_mutex_unlock(&q->lock);
+    mutex_unlock(&q->lock);
+
+    /* Notify other pulling threads that an item can be pulled. */
+    sema_post(&q->s_items);
 
     return true;
 }
 
 bool BPLib_QM_WaitQueueTryPull(BPLib_QM_WaitQueue_t* q, void* ret_item, int timeout_ms)
 {
-    struct timespec deadline;
     int rc;
 
     if ((q == NULL) || (ret_item == NULL))
     {
         return false;
     }
-
-    ms_to_abstimeout((uint32_t)(timeout_ms), &deadline);
-    pthread_mutex_lock(&q->lock);
-    /**** Critical Section Begin ****/
  
     /* Wait for queue to be non-empty */
-    while (q->size == 0)
-    {
-        rc = pthread_cond_timedwait(&q->cv_pull, &q->lock, &deadline);
-        if (rc != 0)
-        {
-            if (rc != ETIMEDOUT)
-            {
-                printf(" BPLib_QM_WaitQueueTryPull NON-TIMEOUT ERROR: %s\n", strerror(rc));
-            }
-            pthread_mutex_unlock(&q->lock);
-            return false;
+    rc = sema_wait_timed_ztimer(&q->s_items, ZTIMER_MSEC, timeout_ms);
+    if (rc != 0) {
+        if (rc != -ETIMEDOUT && rc != -EAGAIN) {
+            printf(" BPLib_QM_WaitQueueTryPull NON-TIMEOUT ERROR: %s\n", strerror(rc));
         }
+        return false;
     }
 
+    mutex_lock(&q->lock);
+    /**** Critical Section Begin ****/
+
     /* Pull an item */
     memcpy(ret_item, (void*)(((char *)q->storage) + (q->front*q->el_size)), q->el_size);
     q->size--;
     q->front = (q->front + 1) % (q->capacity); 
 
-    /* Notify other pushing threads that an item can be pushed */
-    pthread_cond_signal(&q->cv_push);
-
     /**** Critical Section End ****/
-    pthread_mutex_unlock(&q->lock);
+    mutex_unlock(&q->lock);
+
+    /* Notify other pushing threads that an item can be pushed */
+    sema_post(&q->s_slots);
 
     return true;
 }
@@ -194,9 +159,9 @@ bool BPLib_QM_WaitQueueIsEmpty(BPLib_QM_WaitQueue_t* q)
         return false;
     }
 
-    pthread_mutex_lock(&q->lock);
+    mutex_lock(&q->lock);
     IsEmpty = (q->size == 0);
-    pthread_mutex_unlock(&q->lock);
+    mutex_unlock(&q->lock);
 
     return IsEmpty;
 }
@@ -210,9 +175,9 @@ bool BPLib_QM_WaitQueueIsFull(BPLib_QM_WaitQueue_t* q)
         return false;
     }
 
-    pthread_mutex_lock(&q->lock);
+    mutex_lock(&q->lock);
     IsFull = (q->size == q->capacity);
-    pthread_mutex_unlock(&q->lock);
+    mutex_unlock(&q->lock);
 
     return IsFull;
 }
diff --git a/ci/time/inc/bplib_time.h b/ci/time/inc/bplib_time.h
index 5e274ec8..b576ea03 100644
--- a/ci/time/inc/bplib_time.h
+++ b/ci/time/inc/bplib_time.h
@@ -41,7 +41,9 @@
 /**
 ** \brief Time data file name 
 */
+#ifndef BPLIB_TIME_FILE_NAME 
 #define BPLIB_TIME_FILE_NAME "/cf/bplib_time.dat"
+#endif
 
 /**
 ** \brief Length of the time data ring buffers
diff --git a/ci/time/src/bplib_time_internal.c b/ci/time/src/bplib_time_internal.c
index 68f0ba53..e2bcc0be 100644
--- a/ci/time/src/bplib_time_internal.c
+++ b/ci/time/src/bplib_time_internal.c
@@ -27,11 +27,13 @@
 /*
 ** Include
 */
+#include <errno.h>
 
 #include "bplib_time.h"
 #include "bplib_time_internal.h"
 #include "bplib_fwp.h"
 
+#include "vfs_util.h"
 
 /*
 ** Function Definitions
@@ -152,61 +154,33 @@ void BPLib_TIME_SetDtnTimeInBuffer(uint64_t DtnTime, uint32_t BootEra)
 /* Read current time data from file */
 BPLib_Status_t BPLib_TIME_ReadTimeDataFromFile(void)
 {
-    BPLib_Status_t Status = BPLIB_SUCCESS;
-    int32 OsalStatus;
-    
-    /* Try opening existing file or creating new file */
-    OsalStatus = OS_OpenCreate(&BPLib_TIME_GlobalData.FileHandle, BPLIB_TIME_FILE_NAME, 
-                                    OS_FILE_FLAG_CREATE, OS_READ_WRITE);
-    if (OsalStatus == OS_SUCCESS)
-    {
-        /* Read time data from file */
-        OsalStatus = OS_read(BPLib_TIME_GlobalData.FileHandle, (void *) &BPLib_TIME_GlobalData.TimeData, 
+    int res = vfs_file_to_buffer(BPLIB_TIME_FILE_NAME, (void *) &BPLib_TIME_GlobalData.TimeData, 
                                             sizeof(BPLib_TIME_FileData_t));
-
-        /* OSAL should return either 0 (file was just created) or the expected file size */
-        if (OsalStatus != 0 && OsalStatus != sizeof(BPLib_TIME_FileData_t))
-        {
-            Status = BPLIB_TIME_READ_ERROR;
+    if (res != sizeof(BPLib_TIME_FileData_t)) {
+        if (res == -ENOENT) {
+            /* If file not exists try to create it. Reading the just created file back
+             * is not necessary then. */
+            if (BPLib_TIME_WriteTimeDataToFile() == BPLIB_SUCCESS) {
+                return BPLIB_SUCCESS;
+            }
         }
 
-        (void) OS_close(BPLib_TIME_GlobalData.FileHandle);
-    }
-    else 
-    {
-        Status = BPLIB_TIME_READ_ERROR;
+        return BPLIB_TIME_READ_ERROR;
     }
 
-    return Status;
+    return BPLIB_SUCCESS;
 }
 
 /* Write current time data to file */
 BPLib_Status_t BPLib_TIME_WriteTimeDataToFile(void)
 {
-    BPLib_Status_t Status = BPLIB_SUCCESS;
-    int32 OsalStatus;
-
-    /* Try opening existing file or creating new file */
-    OsalStatus = OS_OpenCreate(&BPLib_TIME_GlobalData.FileHandle, BPLIB_TIME_FILE_NAME, 
-                                    OS_FILE_FLAG_CREATE, OS_READ_WRITE);
-    if (OsalStatus == OS_SUCCESS)
-    {
-        /* Dump current time data to file */
-        OsalStatus = OS_write(BPLib_TIME_GlobalData.FileHandle, 
-                (void *) &BPLib_TIME_GlobalData.TimeData, sizeof(BPLib_TIME_FileData_t));
-        if (OsalStatus != sizeof(BPLib_TIME_FileData_t))
-        {
-            Status = BPLIB_TIME_WRITE_ERROR;
-        }
-
-        (void) OS_close(BPLib_TIME_GlobalData.FileHandle);
-    }
-    else 
-    {
-        Status = BPLIB_TIME_WRITE_ERROR;
+    int res = vfs_file_from_buffer(BPLIB_TIME_FILE_NAME, (void *) &BPLib_TIME_GlobalData.TimeData, 
+                                            sizeof(BPLib_TIME_FileData_t));
+    if (res != sizeof(BPLib_TIME_FileData_t)) {
+        return BPLIB_TIME_READ_ERROR;
     }
 
-    return Status;
+    return BPLIB_SUCCESS;
 }
 
 /* Get estimated DTN time */
@@ -251,7 +225,7 @@ uint64_t BPLib_TIME_GetEstimatedDtnTime(BPLib_TIME_MonotonicTime_t MonotonicTime
 int64_t BPLib_TIME_SafeOffset(int64_t HostEpoch, int64_t DtnEpoch, int64_t Multiplier)
 {
     int64_t Offset;
-    int8 Sign;
+    int Sign;
 
     if (HostEpoch < DtnEpoch)
     {
diff --git a/ci/time/src/bplib_time_internal.h b/ci/time/src/bplib_time_internal.h
index e07bf184..481caf4f 100644
--- a/ci/time/src/bplib_time_internal.h
+++ b/ci/time/src/bplib_time_internal.h
@@ -33,7 +33,6 @@
 
 #include "bplib_time.h"
 #include "bplib_fwp.h"
-#include "osapi.h"
 
 
 /*
@@ -89,7 +88,7 @@ typedef struct
 {
     int64_t   CurrentCf;                /**< \brief Most recent CF */
     int64_t   EpochOffset;              /**< \brief Offset between host and DTN epochs */
-    osal_id_t FileHandle;               /**< \brief OSAL handle for time data file */
+    /* FileHandle was removed, is now function local*/
     BPLib_TIME_InitState_t InitState;   /**< \brief Initialization state of TIME */
     BPLib_TIME_FileData_t TimeData;     /**< \brief Local copy of time file data */
 } BPLib_TIME_GlobalData_t;
diff --git a/cla/src/bplib_cla.c b/cla/src/bplib_cla.c
index f78c2782..277f4119 100644
--- a/cla/src/bplib_cla.c
+++ b/cla/src/bplib_cla.c
@@ -443,7 +443,7 @@ BPLib_Status_t BPLib_CLA_ContactTeardown(BPLib_Instance_t *Inst, uint32_t Contac
     }
 
     /* Clear relevant load batch. Ignore return code since we've already done a null check */
-    (void) BPLib_STOR_LoadBatch_Reset(&Inst->BundleStorage.ContactLoadBatches[ContactId]);
+    //(void) BPLib_STOR_LoadBatch_Reset(&Inst->BundleStorage.ContactLoadBatches[ContactId]);
 
     /* Do any framework-specific operations */
     BPLib_FWP_ProxyCallbacks.BPA_CLAP_ContactTeardown(ContactId);
diff --git a/inc/bplib_inst.h b/inc/bplib_inst.h
index 065d0d9f..ca220120 100644
--- a/inc/bplib_inst.h
+++ b/inc/bplib_inst.h
@@ -30,6 +30,7 @@
 #include "bplib_stor.h"
 #include "bplib_pi.h"
 
+#include "mutex.h"
 
 /*
 ** Types
@@ -48,7 +49,7 @@ struct BPLib_Instance
     BPLib_MEM_Pool_t pool; /**< Memory pool for this BPLib Instance */
 
     /* Worker Management */
-    pthread_mutex_t RegisteredWorkersLock; // Move to bplib_os
+    mutex_t RegisteredWorkersLock; // Move to bplib_os
     BPLib_QM_WorkerState_t RegisteredWorkers[QM_MAX_GEN_WORKERS];
     size_t NumWorkers;
 
diff --git a/version/src/bplib_version.c b/version/src/bplib_version.c
index a038c10a..0597efc2 100644
--- a/version/src/bplib_version.c
+++ b/version/src/bplib_version.c
@@ -28,7 +28,7 @@
 *************************************************************************/
 #include "bplib_api_types.h"
 #include "bplib_version.h"
-#include "osapi.h" /* for OS_Printf() */
+/* #include "osapi.h" removed, for OS_Printf() */
 #include <stdio.h> /* for "snprintf()" */
 
 /*************************************************************************
@@ -44,15 +44,12 @@
 /* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
 int BPLib_PrintVersion(void)
 {
-    char VersionString[BPLIB_CFG_MAX_VERSION_STR_LEN];
-
-    (void) snprintf(VersionString, BPLIB_CFG_MAX_VERSION_STR_LEN, "v%u.%u.%u-sprint-%u",
+    /* Print the version using printf, which should be provided by RIOT */
+    (void) printf("BPLib Version: v%u.%u.%u-sprint-%u\n",
         BPLIB_MAJOR_VERSION,
         BPLIB_MINOR_VERSION,
         BPLIB_REVISION,
         BPLIB_BUILD_NUMBER);
 
-    OS_printf("BPLib Version: %s\n", VersionString);
-
     return BPLIB_SUCCESS;
 }
-- 
2.43.0

